### PR TITLE
[JAX] Flatten_axis for quantization and Sharding propagation fixes

### PR DESCRIPTION
**Pull request by phu0ngng**
# Description

In #1627, we enforced all the tensors to be flattenable to 2D tensor with axis=-1. This requires additional reshaping in JAX that merges dimensions, resulting in the loss of sharding information. 

In this PR, we introduced `flatten_axis` that allows flattening the tensor to 2D via any axis. With this, merging axes is no longer needed; thus, the sharding information can be propagated correctly. 

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [x] Code refactoring

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/jax/cpp_extensions/misc.py [hunk 1]
```diff
@@ -224,23 +224,29 @@ def try_apply_delayed_scaling_2x_war(f, *args, quantizer=None, **kwargs):
...
     other_outputs = None
     if isinstance(rowwise, tuple):
         other_outputs = rowwise[1:]
         rowwise = rowwise[0]
-    quantizer.q_axis = QuantizeAxis.ROWWISE_COLWISE
-    colwise_data = jnp.transpose(rowwise.data, (-1, *range(rowwise.data.ndim - 1)))
+    quantizer.q_layout = QuantizeLayout.ROWWISE_COLWISE
+    if flatten_axis < 0:
+        flatten_axis += rowwise.data.ndim
+    assert 0 < flatten_axis < rowwise.data.ndim, "flatten_axis is out of bounds"
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/cpp_extensions/misc.py:236]: This is `0 < flatten_axis` rather than `0 <= flatten_axis` as we need at least one axis to the left of the flatten axis so after flattening we have 2 axes to give to TE, right?

File: transformer_engine/jax/csrc/extensions/activation.cpp [hunk 2]
```diff
@@ -58,24 +52,47 @@ Error_Type ActLuFFI(cudaStream_t stream, Buffer_Type input_buf, Buffer_Type scal
         std::vector<size_t>{
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/csrc/extensions/activation.cpp:52]: I don't think we need this call for `set_rowwise_scale_inv` anymore here as we set it for both scaling modes below, right? Maybe we could remove or replace it with an NVTE_CHECK for supported scaling modes until we support current scaling

File: transformer_engine/jax/csrc/extensions/activation.cpp [hunk 3]
```diff
@@ -190,9 +209,9 @@ pybind11::tuple GetDActDBiasQuantizeWorkspaceSizes(size_t batch_size, size_t hid
 
 Error_Type DActLuDBiasQuantizeFFI(cudaStream_t stream, Buffer_Type input_buf,
                                   Buffer_Type act_input_buf, Buffer_Type scale_buf,
-                                  Result_Type output_buf, Result_Type output_trans_buf,
-                                  Result_Type scale_inv_buf, Result_Type trans_scale_inv_buf,
-                                  Result_Type amax_out_buf, Result_Type dbias_buf,
+                                  Result_Type output_buf, Result_Type colwise_output_buf,
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/csrc/extensions/activation.cpp:212]: Thanks for updating this code to be consistent with our other naming! I think I introduced that inconsistent naming originally

**Review by jberchtold-nvidia — APPROVED** [review]

File: transformer_engine/jax/cpp_extensions/misc.py (hunk 1 above)
Comment (phu0ngng) [reply transformer_engine/jax/cpp_extensions/misc.py:236]: Yes, TE requires the data to be 2D for MXFP8. So we can't accept axis = 0 and mistakenly flatten the data to 1D.

File: transformer_engine/jax/csrc/extensions/activation.cpp (hunk 2 above)
Comment (phu0ngng) [reply transformer_engine/jax/csrc/extensions/activation.cpp:52]: Agree. Same for the DActDBiasQuantizeFFI.
Removed.

File: transformer_engine/jax/quantize/scaling_modes.py [hunk 4]
```diff
@@ -122,16 +122,25 @@ def get_scale_dtype(self) -> jnp.dtype:
...
+            assert n_scale_blocks % last == 0
+            n_scale_blocks //= last
+            # handle middle dim, exclude first and last
+            for mid in reversed(data_shape[1:-1]):
+                scale_shape = (mid,) + scale_shape
+                assert n_scale_blocks % mid == 0
+                n_scale_blocks //= mid
+            scale_shape = (n_scale_blocks,) + scale_shape
+        else:
+            scale_shape = (n_scale_blocks,)
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/quantize/scaling_modes.py:141]: Do we support quantizing 1D tensors at all?

Comment (phu0ngng) [reply transformer_engine/jax/quantize/scaling_modes.py:141]: No, we don't. 

But this function gets called for each part of the 2D-flattened tensor here https://github.com/NVIDIA/TransformerEngine/pull/1644/files#diff-0158638e30529db0bb268ae65eb085b2d22b52e6e0ff4891fe2c7ea9959eea79R209-R214 
Therefore, the `datashape` can be 1D. Perhaps I should rename it to `partial_data_shape`.

Comment (jberchtold-nvidia) [reply transformer_engine/jax/quantize/scaling_modes.py:141]: Oh I see now, that makes sense. Thanks!

File: transformer_engine/jax/cpp_extensions/activation.py [hunk 5]
```diff
@@ -670,15 +665,16 @@ def infer_sharding_from_operands(
...
         x_spec = get_padded_spec(arg_infos[1])
+        scale_spec = get_padded_spec(arg_infos[2])
 
         out_sharding = NamedSharding(
             mesh, PartitionSpec(*x_spec), desc="DActLuDBiasQuantizePrimitive.out"
         )
         if is_2x:
             if scaling_mode == ScalingMode.NVTE_DELAYED_TENSOR_SCALING.value:
-                colwise_x_spec = multidim_transpose(x_spec)
+                colwise_x_spec = multidim_transpose(x_spec, transpose_axis=-2)
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/cpp_extensions/activation.py:677]: Note for @jreiffers, change in signature of this function

File: transformer_engine/jax/cpp_extensions/activation.py [hunk 6]
```diff
@@ -340,36 +333,40 @@ def partition(
     ):
         del result_infos, is_outer  # Unused.
         x_spec = get_padded_spec(arg_infos[0])
-        out_spec = (*x_spec[:-1], x_spec[-1])
-        if act_len == 2 and x_spec[-1] is None:
-            # Ensure last axis is partitioned and not the gating axis
-            out_spec = (*x_spec[:-2], None, x_spec[-2])
+        scale_spec = get_padded_spec(arg_infos[1])
+
+        out_spec = (*x_spec[:-2], x_spec[-1])
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/cpp_extensions/activation.py:338]: Note for @jreiffers, how the signature of this primitive changed in this PR
